internal/store/inmemory/inmemory.go

```diff
@@ -0,0 +1,227 @@
+		return errors.New(
+			fmt.Sprintf(
+				"Error upserting route - id = %s, svc = %s, deploy = %s, ns = %s, domains = %s - must not be empty",
+				id, service, deploy, namespace, domains),
+		)
+	}
+
+	// error if deployment or domains are already associated to another route
+	err := s.checkDeployAndDomainsOwnership(id, deploy, namespace, domains)
+
+	if err != nil {
+		return err
+	}
+
+	if existingRoute, ok := s.m[id]; ok {
+		if port == "" {
+			return errors.New(
+				fmt.Sprintf(
+					"Error updating route - port = %s must not be empty", port),
+			)
+		}
+
+		// /!\ this need to be on top - otherwise the data will have already been overriden in the route
+		newKeys := s.cleanStore(
+			existingRoute.GetDeployment(), existingRoute.GetNamespace(), existingRoute.GetDomains(),
+			deploy, namespace, domains)
+
+		// associate the route to new deployment key / domains
+		for _, k := range newKeys {
+			s.m[k] = existingRoute
+		}
+
+		// TODO check the errors
+		_ = existingRoute.SetService(service)
+		_ = existingRoute.SetPort(port)
+		_ = existingRoute.SetDeployment(deploy)
+		_ = existingRoute.SetDomains(domains)
+		// route is a pointer and it's changing dynamically - no need to "persist" the change in the map
+	} else {
+		newRoute, err := model.NewRoute(id, service, port, deploy, namespace, domains)
+
+		if err != nil {
+			return err
+		}
+
+		s.createRoute(newRoute)
+	}
+
+	return nil
+}
+
+// return an error if deploy or domains are already associated to a different id
+func (s *inMemoryStore) checkDeployAndDomainsOwnership(id, deploy, ns string, domains []string) error {
+	r, err := s.GetRouteByDeployment(deploy, ns)
+
+	if err == nil && r.GetId() != id {
+		return errors.New(fmt.Sprintf("Deployment %s.%s is already owned by %s", deploy, ns, r.GetId()))
+	}
+
+	for _, d := range domains {
+		r, err = s.GetRouteByDomain(d)
+
+		if err == nil && r.GetId() != id {
+			return errors.New(fmt.Sprintf("Domain %s is already owned by %s", d, r.GetId()))
+		}
+	}
+
+	return nil
+}
+
+func (s *inMemoryStore) createRoute(route *model.Route) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	deploymentKey := genDeploymentKey(route.GetDeployment(), route.GetNamespace())
+	s.m[route.GetId()] = route
+	s.m[deploymentKey] = route
+	for _, d := range route.GetDomains() {
+		s.m[d] = route
+	}
+}
+
+// Remove old domains and deployment from the store if they are not == new ones
+func (s *inMemoryStore) cleanStore(
+	oldDeploy, oldNs string, oldDomains []string,
+	newDeploy, newNs string, newDomains []string) []string {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	var newKeys []string
+
+	deployKeyNotInStore := s.cleanOldDeploymentFromStore(oldDeploy, oldNs, newDeploy, newNs)
+
+	if deployKeyNotInStore != "" {
+		newKeys = append(newKeys, deployKeyNotInStore)
+	}
+
+	domainsNotInStore := s.cleanOldDomainsFromStore(oldDomains, newDomains)
+
+	if newDomains != nil {
+		newKeys = append(newKeys, domainsNotInStore...)
```

Comment:
comment - could just return here if you initialize newKeys := []string on line 115.